lib/graphql/stitching/shaper.rb

```diff
@@ -0,0 +1,84 @@
+# typed: false
+# frozen_string_literal: true
+
+module GraphQL
+  module Stitching
+    class Shaper
+      def initialize(supergraph:, document:)
+        @supergraph = supergraph
+        @document = document
+      end
+
+      def perform(raw_result)
+        if ! raw_result.key?("errors") || raw_result["errors"].empty?  # TODO should do null cleanup for errors too
+          munge_entry(raw_result["data"], @document.operation.selections, @supergraph.schema.query)
+        end
+
+        # hate doing a second pass, but cannot remove _STITCH_ fields until the fragements are processed
+        clean_entry(raw_result["data"])
+
+        raw_result
+      end
+
+      private
+
+      def munge_entry(entry, selections, parent_type)
+        selections.each do |node|
+          if node.respond_to?(:name) && node&.name == "__typename"
+            next
+          end
+
+          case node
+          when GraphQL::Language::Nodes::Field
+            munge_field(entry, node, parent_type)
+
+          when GraphQL::Language::Nodes::InlineFragment
+            next unless entry["_STITCH_typename"] == node.type.name
+            fragment_type = @supergraph.schema.types[node.type.name]
+            munge_entry(entry, node.selections, fragment_type)
+
+          when GraphQL::Language::Nodes::FragmentSpread
+            next unless entry["_STITCH_typename"] == node.name
+            fragment = @document.fragment_definitions[node.name]
+            fragment_type = @supergraph.schema.types[node.name]
+            munge_entry(entry, fragment.selections, fragment_type)
+          else
+            raise "Unexpected node of type #{node.class.name} in selection set."
+          end
+        end
+      end
+
+      def munge_field(entry, node, parent_type)
+        field_identifier = (node.alias || node.name)
+        node_type = Util.get_field_type(@supergraph, parent_type, node)
+
+        child_entry = entry[field_identifier]
+        if child_entry.nil?
+          entry[field_identifier] = nil
+        elsif child_entry.is_a? Array
+          child_entry.each do |raw_item|
+            munge_entry(raw_item, node.selections, node_type)
+          end
+        elsif ! Util.is_leaf_type?(node_type)
+          munge_entry(child_entry, node.selections, node_type)
+        end
+      end
+
+      def clean_entry(entry)
+        return if entry.nil?
+
+        entry.each do |key, value|
```

Comment:
I _think_ we can just do this as the first thing that happens in `munge_entry`...

```ruby
def munge_entry(entry, selections, parent_type)
  entry.reject! { |k, v| k..start_with? "_STITCH_" }
  # ...
```

Seems like that should make it recursive while not needing to traverse the tree twice.

Reply:
I love the way that looks, but every time I have tried it fragments have failed to resolve correctly. Now that things are working more consistently, let me pin down exactly what goes missing on the fragements that prevents them from resolving

Reply:
let's consider revisiting the optimization after we get null bubbling working?

1. make it work
2. make it work right
3. make it work right, fast, and pretty

Reply:
OH!!! Of course... fragments would fail because they need `_STITCH_typename` for type inspection. I believe it would work if that rejection line moved to the bottom of `munge_entry` as the last operation performed on each composite scope visited.